### ISSUE TITLE
Understanding post-hoc explanations on b-cos ViTs

### ISSUE THREAD

**Issue by rushin682** [B-cos/B-cos-v2#4]
Hey Moritz,

Thanks for this exciting work. Congratulations :)

Considering that weight-input alignments can make the networks inherently interpretable (which makes total sense), **I would think that the post-hoc explanations then benefit from this alignment too.** 

Specifically referring to the widely used Chefer et al. LRP approach **which can also highlight class-specific heatmaps**, it would be nice to know your insights on whether the explanations from b-cos networks highlight the same regions if plotting the heat maps using the explanation module from the repo vs the LRP approach. Were you able to perform those experiments too? 

As far as I understand from the paper, the experiments for LRP were only performed on conventional ViTs. I was curious if you tested them on the b-cos ViTs too...

Thanks once again for the awesome research. I look forward to your reply.

best,
Rushin

**Comment by moboehle** [comment]
Hi Rushin,

thank you for your interest! Your intuition seems to be correct — e.g., as we show in Fig. 6 in [the paper](https://arxiv.org/abs/2306.10898), post-hoc explanation methods indeed seem to yield better results for the B-cos version of a given conventional model (compare row 1+3 and 2+4). We observe similar behaviour when evaluating Integrated Gradients on B-cos ViTs vs. standard ViTs (see Table 4).

Unfortunately, the method by Chefer et al. relies on LRP, which requires propagation rules to be defined for each layer, which have not been developed in the context of LRP for B-cos transformations yet. That said, it is certainly an interesting direction to further investigate. However, two points need to be kept in mind here: first, a direct comparison between Chefer's method on standard ViTs and B-cos ViTs is not trivial, as the activation functions (GELU vs. Identity) and layer-wise transformations (B-cos Conv2d vs. linear Conv2d) differ. As a result, the individual effect of the chosen LRP rules, the layer-wise transformations, and the activation functions are difficult to disentangle. 

Secondly, as Chefer et al. state in their introduction, their primary goal is to derive mechanistic heuristics to improve on interpretability benchmarks; specifically, they show that a particular combination of LRP and model gradients yield particularly good results for the ViT model under investigation. Most likely, the B-cos explanations can be modified in a similar fashion for improved performance on downstream tasks such as segmentation and localisation. While this can be advantageous in certain settings, it should be noted that this would come at the cost of model-faithfulness; whether or not this trade-off is acceptable will eventually depend on user preference. 

I hope you find this answer helpful!

Best,
Moritz

**Comment by rushin682** [comment]
Hi Moritz, 

Thanks for the quick reply. 
The following paper by Chefer et al. (MM-explainibility) removed the LRP rules for each layer since they are computationally expensive. Accordingly now they use gradient propogations for computing the relevance of each transformer layer. I tried using bcos layers for ff networks as proposed in the repository and the results for the Chefer's explanations seem to improve. 

I also was curious regarding multi-class classification problems and how the concept of aligning weights to specific classes works in such scenarios. If I remember correctly, all experiments were done on BCE loss. What suggestions would you give in applying your framework for multi-class/multi-label problems and how to conceptually understand the alignment in such cases?

Thanks again for your the research direction :)

best,
Rushin

**Comment by moboehle** [comment]
Hi Rushin,

thank you for clarifying which paper by Chefer et al. you were referring to 🙂 

Regarding the multi-class classification: please note that while we compute the loss for each class individually via the BCE loss, we still train our models in the standard multi-class classification setup. I.e., instead of using the CE loss with softmax, we train the B-cos models with a sigmoid activation function and the BCE loss per class.

Best,
Moritz
